docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+
+The part of the host-side API which deals with device management and their
+queries are synchronous. All asynchronous APIs, such as kernel execution, data
+movement and potentially data allocation/freeing all happen in the context of
+device streams.
+
+Streams are FIFO buffers of commands to execute relating to a given device.
+Commands which enqueue tasks on a stream all return promptly and the command is
+executed asynchronously. All side-effects of a command on a stream are visible
+to all subsequent commands on the same stream. Multiple streams may point to
+the same device and those streams may be fed from multiple concurrent host-side
+threads. Execution on multiple streams may be concurrent but isn't required to
+be.
+
+Asynchronous APIs involving a stream all return a stream event which may be
+used to synchronize the execution of multiple streams. A user may enqueue a
+barrier onto a stream referencing an event. The barrier will will block until
+the command related to the event does not complete, at which point all
+side-effects of the command shall be visible to commands following the barrier,
+even if those side-effects manifest on different devices.
+
+Streams also support executing user-defined functions as callbacks on the host.
+The stream will not launch subsequent commands until the callback completes.
+
+Device-side execution
+-------------------------------------------------------------------------------
+
+The SIMT programming model behind the HIP device-side execution is a
+middle-ground between SMT (Simultaneous Multi-Threading) programming known from
+multi-core CPUs, and SIMD (Single Instruction, Multiple Data) programming
+mostly known from exploiting relevant instruction sets on CPUs (eg.
+SSE/AVX/Neon).
+
+A HIP device compiler maps our SIMT code written in HIP C++ to an inherently
+SIMD architecture (like GPUs) not by exploiting data parallelism within a
+single instance of a kernel and spreading identical instructions over the SIMD
+engines at hand, but by scalarizing the entire kernel and issuing the scalar
+instructions of multiple kernel instances to each of the SIMD engine lanes.
+
+Kernel launch
+-------------------------------------------------------------------------------
+
+Kernels may be launched in multiple ways all with different syntaxes and
+intended use-cases.
+
+* Using the triple-chevron ``<<<...>>>`` operator on a ``__global__`` annotated
+  function.
+
+* Using ``hipLaunchKernelGGL()`` on a ``__global__`` annotated function.
+
+  .. tip::
+
+    This name by default is a macro expanding to triple-chevron. In cases where
+    language syntax extensions are undesirable, or where launching templated
+    and/or overloaded kernel functions define the
+    ``HIP_TEMPLATE_KERNEL_LAUNCH`` preprocessor macro before including the HIP
+    headers to turn it into a templated function.
+
+* Using the
+  :doxygen:`launch APIs supporting the triple-chevron syntax <Clang>` directly.
```

Comment:
This does not seem as if it's supposed to look this way. Especially in the rendered version.

Reply:
Couldn't fix it either, not even sure what the original target of this was - removed the reference